The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9115078933/job/25060554126) [(plain)](https://github.com/rust-lang/rust/commit/f55ef48d9edf7af00ad743e4d8580210fef706fe/checks/25060554126/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling build_helper v0.1.0 (/checkout/src/tools/build_helper)
error[E0061]: this method takes 1 argument but 0 arguments were supplied
##[error]    --> src/core/config/config.rs:1732:41
     |
1732 |             config.channel = ci_channel.clone_into();
     |                                         ^^^^^^^^^^-- an argument of type `&mut std::string::String` is missing
note: method defined here
    --> /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/borrow.rs:77:8
help: provide the argument
     |
     |
1732 |             config.channel = ci_channel.clone_into(/* &mut std::string::String */);

error[E0308]: mismatched types
##[error]    --> src/core/config/config.rs:1732:30
     |
     |
1732 |             config.channel = ci_channel.clone_into();
     |             --------------   ^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
     |             expected due to the type of this binding

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0061]: this method takes 1 argument but 0 arguments were supplied
##[error]    --> src/core/config/config.rs:1732:41
     |
1732 |             config.channel = ci_channel.clone_into();
     |                                         ^^^^^^^^^^-- an argument of type `&mut std::string::String` is missing
note: method defined here
    --> /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/borrow.rs:77:8
help: provide the argument
     |
     |
1732 |             config.channel = ci_channel.clone_into(/* &mut std::string::String */);

error[E0308]: mismatched types
##[error]    --> src/core/config/config.rs:1732:30
     |
     |
1732 |             config.channel = ci_channel.clone_into();
     |             --------------   ^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
     |             expected due to the type of this binding

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0061]: this method takes 1 argument but 0 arguments were supplied
##[error]    --> src/core/config/config.rs:1732:41
     |
1732 |             config.channel = ci_channel.clone_into();
     |                                         ^^^^^^^^^^-- an argument of type `&mut std::string::String` is missing
note: method defined here
    --> /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/borrow.rs:77:8
help: provide the argument
     |
     |
1732 |             config.channel = ci_channel.clone_into(/* &mut std::string::String */);

error[E0308]: mismatched types
##[error]    --> src/core/config/config.rs:1732:30
     |
     |
1732 |             config.channel = ci_channel.clone_into();
     |             --------------   ^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
     |             expected due to the type of this binding

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0061]: this method takes 1 argument but 0 arguments were supplied
##[error]    --> src/core/config/config.rs:1732:41
     |
1732 |             config.channel = ci_channel.clone_into();
     |                                         ^^^^^^^^^^-- an argument of type `&mut std::string::String` is missing
note: method defined here
    --> /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/borrow.rs:77:8
help: provide the argument
     |
     |
1732 |             config.channel = ci_channel.clone_into(/* &mut std::string::String */);

error[E0308]: mismatched types
##[error]    --> src/core/config/config.rs:1732:30
     |
     |
1732 |             config.channel = ci_channel.clone_into();
     |             --------------   ^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
     |             expected due to the type of this binding

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0061]: this method takes 1 argument but 0 arguments were supplied
##[error]    --> src/core/config/config.rs:1732:41
     |
1732 |             config.channel = ci_channel.clone_into();
     |                                         ^^^^^^^^^^-- an argument of type `&mut std::string::String` is missing
note: method defined here
    --> /rustc/6b544f5ff8d45221d61962651a5f5ab9fe535e16/library/alloc/src/borrow.rs:77:8
help: provide the argument
     |
     |
1732 |             config.channel = ci_channel.clone_into(/* &mut std::string::String */);

error[E0308]: mismatched types
##[error]    --> src/core/config/config.rs:1732:30
     |
     |
1732 |             config.channel = ci_channel.clone_into();
     |             --------------   ^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
     |             expected due to the type of this binding

Some errors have detailed explanations: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
```

</details>

--- a/src/bootstrap/src/core/config/config.rs
+++ b/src/bootstrap/src/core/config/config.rs
@@ -1307,6 +1307,9 @@ impl Config {
             toml_path = config.src.join(toml_path);
         }
 
+        let file_content = t!(fs::read_to_string(config.src.join("src/ci/channel")));
+        let ci_channel = file_content.trim_end();
+
         // Give a hard error if `--config` or `RUST_BOOTSTRAP_CONFIG` are set to a missing path,
         // but not if `config.toml` hasn't been created.
         let mut toml = if !using_default_path || toml_path.exists() {
@@ -1533,6 +1536,7 @@ impl Config {
         let mut optimize = None;
         let mut omit_git_hash = None;
 
+        let mut is_user_configured_rust_channel = false;
         if let Some(rust) = toml.rust {
             let Rust {
                 optimize: optimize_toml,
@@ -1590,6 +1594,7 @@ impl Config {
                 lld_mode,
             } = rust;
 
+            is_user_configured_rust_channel = channel.is_some();
             set(&mut config.channel, channel);
 
             config.download_rustc_commit = config.download_ci_rustc_commit(download_rustc);
@@ -1597,8 +1602,6 @@ impl Config {
             if config.download_rustc_commit.is_some() {
                 // We need the channel used by the downloaded compiler to match the one we set for rustdoc;
                 // otherwise rustdoc-ui tests break.
-                let ci_channel = t!(fs::read_to_string(config.src.join("src/ci/channel")));
-                let ci_channel = ci_channel.trim_end();
                 if config.channel != ci_channel
                     && !(config.channel == "dev" && ci_channel == "nightly")
                 {
@@ -1724,6 +1727,10 @@ impl Config {
         let default = config.channel == "dev";
         config.omit_git_hash = omit_git_hash.unwrap_or(default);
         config.rust_info = GitInfo::new(config.omit_git_hash, &config.src);
+
+        if config.rust_info.is_from_tarball() && !is_user_configured_rust_channel {
+            ci_channel.clone_into(&mut config.channel);
+        }
 
         if let Some(llvm) = toml.llvm {
             let Llvm {

